llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -12982,7 +12982,15 @@ InstructionCost BoUpSLP::getGatherCost(ArrayRef<Value *> VL, bool ForPoisonSrc,
               TTI::SK_InsertSubvector, VecTy, std::nullopt, CostKind,
               I * ScalarTyNumElements, cast<FixedVectorType>(ScalarTy));
     } else {
-      Cost = TTI->getScalarizationOverhead(VecTy, ~ShuffledElements,
+      // Add insertion costs for all elements, but not for loads that can be
+      // loaded directly into a vector element for free.
+      APInt FreeEltLoads = APInt::getZero(VL.size());
+      if (TTI->supportsEfficientVectorElementLoadStore())
+        for (unsigned I : seq<unsigned>(VL.size()))
+          if (VL[I]->hasOneUse() && isa<LoadInst>(VL[I]))
+            FreeEltLoads.setBit(I);
+      APInt DemandedElts = ~ShuffledElements & ~FreeEltLoads;
+      Cost = TTI->getScalarizationOverhead(VecTy, DemandedElts,
```

Comment:
I'm afraid this may affect the targets, which do not support it. Instead better to add an extra parameter for this function to account free loads.

Reply:
How could it affect other targets? Please clarify. AFIK, the FreeEltLoads is 0, so if a target does not support efficient vector element loads, it remains 0, so later on it becomes all-ones after inversion and is therefore a no-op in the bitwise AND (&) operation. So it seems in this case we pass in ~ShuffleElements just like before...

Reply:
Ah, missed that it is guarded by a check